src/hayhooks/server/pipelines/utils.py

```diff
@@ -121,57 +244,153 @@ async def async_streaming_generator(
+async def _execute_pipeline_async(
+    pipeline: Union[Pipeline, AsyncPipeline, Agent], pipeline_run_args: Dict[str, Any]
+) -> asyncio.Task:
+    """
+    Creates and returns an async task for pipeline execution.
 
-    pipeline_run_args = pipeline_run_args.copy()
+    Args:
+        pipeline: The pipeline or agent to execute
+        pipeline_run_args: Execution arguments
 
-    if streaming_component_name not in pipeline_run_args:
-        pipeline_run_args[streaming_component_name] = {}
+    Returns:
+        Async task for pipeline execution
+    """
+    if isinstance(pipeline, AsyncPipeline):
+        return asyncio.create_task(pipeline.run_async(data=pipeline_run_args))
+    elif isinstance(pipeline, Agent):
+        return asyncio.create_task(pipeline.run_async(**pipeline_run_args))
+    else:  # Regular Pipeline
+        return asyncio.create_task(asyncio.to_thread(pipeline.run, data=pipeline_run_args))
 
-    streaming_component.streaming_callback = streaming_callback
 
-    if isinstance(pipeline, AsyncPipeline):
-        pipeline_task = asyncio.create_task(pipeline.run_async(data=pipeline_run_args))
-    else:
-        pipeline_task = asyncio.create_task(asyncio.to_thread(pipeline.run, data=pipeline_run_args))
+async def _stream_chunks_from_queue(
+    queue: asyncio.Queue[StreamingChunk], pipeline_task: asyncio.Task
+) -> AsyncGenerator[StreamingChunk, None]:
+    """
+    Streams chunks from the queue while the pipeline is running.
+
+    Args:
+        queue: Queue containing streaming chunks
+        pipeline_task: The async task running the pipeline
+
+    Yields:
+        StreamingChunk: Individual chunks from the pipeline
+    """
+    while not pipeline_task.done() or not queue.empty():
+        # Check for pipeline completion with exception
+        if pipeline_task.done():
+            exception = pipeline_task.exception()
+            if exception is not None:
+                raise exception
+
+        try:
+            chunk = await asyncio.wait_for(queue.get(), timeout=0.1)
+            yield chunk
+        except asyncio.TimeoutError:
+            continue
+        except asyncio.CancelledError:
+            log.warning("Async streaming generator was cancelled")
+            break
+        except Exception as e:
+            log.error(f"Unexpected error in async streaming generator: {e}")
+            raise e
+
+
+async def _cleanup_pipeline_task(pipeline_task: asyncio.Task) -> None:
+    """
+    Cleans up the pipeline task if it's still running.
+
+    Args:
+        pipeline_task: The task to clean up
+    """
+    if not pipeline_task.done():
+        pipeline_task.cancel()
+        try:
+            await asyncio.wait_for(pipeline_task, timeout=1.0)
+        except (asyncio.TimeoutError, asyncio.CancelledError):
+            pass
+        except Exception as e:
+            log.warning(f"Error during pipeline task cleanup: {e}")
+            raise e
+
+
+async def async_streaming_generator(
```

Comment:
Public API, let's add more docs, example, how everything works etc etc

Reply:
Will do in the next PR as stated in description ;)